### ISSUE TITLE
DyanomoDB Unmarshal

### ISSUE THREAD

**Issue by rjocoleman** [apex/go-apex#17]
Below is the example JSON event from the Lambda management console for DynamoDB stream events.

Currently the unmarshal for this is coupled to some JSON output in the dynamo docs: https://github.com/apex/go-apex/blob/master/dynamo/dynamo.go#L23-L30
`ForumName` and `Subject` are dynamic, they're the key names.  In the below example the keys are `Message` and `Id`. 
Additional they JSON key name is significant `N` or `S` in this example, that's the dynamo [AttributeValue](http://docs.aws.amazon.com/dynamodbstreams/latest/APIReference/API_AttributeValue.html) - here `number` or `string` respectively. 

Dynamo Events also support 4 modes, the below example is 

```
KEYS_ONLY - only the key attributes of the modified item
NEW_IMAGE - the entire item, as it appears after it was modified
OLD_IMAGE - the entire item, as it appeared before it was modified
NEW_AND_OLD_IMAGES- both the new and the old images of the item
```

I haven't 

``` json
{
  "Records": [
    {
      "eventID": "1",
      "eventVersion": "1.0",
      "dynamodb": {
        "Keys": {
          "Id": {
            "N": "101"
          }
        },
        "NewImage": {
          "Message": {
            "S": "New item!"
          },
          "Id": {
            "N": "101"
          }
        },
        "StreamViewType": "NEW_AND_OLD_IMAGES",
        "SequenceNumber": "111",
        "SizeBytes": 26
      },
      "awsRegion": "us-west-2",
      "eventName": "INSERT",
      "eventSourceARN": "arn:aws:dynamodb:us-west-2:account-id:table/ExampleTableWithStream/stream/2015-06-27T00:48:05.899",
      "eventSource": "aws:dynamodb"
    },
    {
      "eventID": "2",
      "eventVersion": "1.0",
      "dynamodb": {
        "OldImage": {
          "Message": {
            "S": "New item!"
          },
          "Id": {
            "N": "101"
          }
        },
        "SequenceNumber": "222",
        "Keys": {
          "Id": {
            "N": "101"
          }
        },
        "SizeBytes": 59,
        "NewImage": {
          "Message": {
            "S": "This item has changed"
          },
          "Id": {
            "N": "101"
          }
        },
        "StreamViewType": "NEW_AND_OLD_IMAGES"
      },
      "awsRegion": "us-west-2",
      "eventName": "MODIFY",
      "eventSourceARN": "arn:aws:dynamodb:us-west-2:account-id:table/ExampleTableWithStream/stream/2015-06-27T00:48:05.899",
      "eventSource": "aws:dynamodb"
    },
    {
      "eventID": "3",
      "eventVersion": "1.0",
      "dynamodb": {
        "Keys": {
          "Id": {
            "N": "101"
          }
        },
        "SizeBytes": 38,
        "SequenceNumber": "333",
        "OldImage": {
          "Message": {
            "S": "This item has changed"
          },
          "Id": {
            "N": "101"
          }
        },
        "StreamViewType": "NEW_AND_OLD_IMAGES"
      },
      "awsRegion": "us-west-2",
      "eventName": "REMOVE",
      "eventSourceARN": "arn:aws:dynamodb:us-west-2:account-id:table/ExampleTableWithStream/stream/2015-06-27T00:48:05.899",
      "eventSource": "aws:dynamodb"
    }
  ]
}
```

I'm not sure how, to deal with this... but Dynamo won't work as is. Perhaps something like this:

``` diff
diff --git a/dynamo/dynamo.go b/dynamo/dynamo.go
index e32bef1..771abf3 100644
--- a/dynamo/dynamo.go
+++ b/dynamo/dynamo.go
@@ -12,6 +12,9 @@ type Event struct {
        Records []*Record `json:"Records"`
 }

+// StreamRecord represents the Dynamo keys
+type StreamRecord interface{}
+
 // Record represents a single Dynamo record.
 type Record struct {
        EventID      string `json:"eventID"`
@@ -20,14 +23,9 @@ type Record struct {
        EventVersion string `json:"eventVersion"`
        AWSRegion    string `json:"awsRegion"`
        Dynamodb     struct {
-               Keys struct {
-                       ForumName struct {
-                               S string `json:"S"`
-                       } `json:"ForumName"`
-                       Subject struct {
-                               S string `json:"S"`
-                       } `json:"Subject"`
-               } `json:"Keys"`
+               Keys           StreamRecord
+               NewImage       StreamRecord
+               OldImage       StreamRecord
                SequenceNumber string `json:"SequenceNumber"`
                SizeBytes      int    `json:"SizeBytes"`
                StreamViewType string `json:"StreamViewType"`
```

**Comment by rjocoleman** [comment]
Another, probably better, option would be to use another lib to unmarshal. e.g. the AWS Go SDK:

``` diff
diff --git a/dynamo/dynamo.go b/dynamo/dynamo.go
index e32bef1..a92e354 100644
--- a/dynamo/dynamo.go
+++ b/dynamo/dynamo.go
@@ -5,6 +5,7 @@ import (
    "encoding/json"

    "github.com/apex/go-apex"
+   "github.com/aws/aws-sdk-go/service/dynamodbstreams"
 )

 // Event represents a Dynamo event with one or more records.
@@ -14,24 +15,12 @@ type Event struct {

 // Record represents a single Dynamo record.
 type Record struct {
-   EventID      string `json:"eventID"`
-   EventName    string `json:"eventName"`
-   EventSource  string `json:"eventSource"`
-   EventVersion string `json:"eventVersion"`
-   AWSRegion    string `json:"awsRegion"`
-   Dynamodb     struct {
-       Keys struct {
-           ForumName struct {
-               S string `json:"S"`
-           } `json:"ForumName"`
-           Subject struct {
-               S string `json:"S"`
-           } `json:"Subject"`
-       } `json:"Keys"`
-       SequenceNumber string `json:"SequenceNumber"`
-       SizeBytes      int    `json:"SizeBytes"`
-       StreamViewType string `json:"StreamViewType"`
-   } `json:"dynamodb"`
+   EventID      string                        `json:"eventID"`
+   EventName    string                        `json:"eventName"`
+   EventSource  string                        `json:"eventSource"`
+   EventVersion string                        `json:"eventVersion"`
+   AWSRegion    string                        `json:"awsRegion"`
+   Dynamodb     *dynamodbstreams.StreamRecord `json:"dynamodb"`
 }

 // Handler handles Dynamo events.
```

Which is usable like so:

``` go
func main() {
    apexDynamo.HandleFunc(func(event *apexDynamo.Event, ctx *apex.Context) error {
        for _, record := range event.Records {
            if record.EventName == "INSERT" {
                fmt.Println(*record.Dynamodb.NewImage["foo"].S)
            }
        }

        return nil
    })
}
```

**Comment by tj** [comment]
Oops, guess I shouldn't copy/paste the json from an example. Using the SDK sounds fine to me, though they usually use pointers unconventionally, might be better to just copy those struct(s) in

**Comment by rjocoleman** [comment]
There's a lot going on in the DynamoDB Attribute package for converting and marshaling to actually make it useful, it's more complex than just a few structs unfortunately as the data comes back from the API in a convoluted way!

I'll finish up this integration I'm working on, it's got several Dynamo touch points, then I'll send an appropriate PR for consideration and discussion.

**Comment by harlow** [comment]
I've been looking at the DynamoDB streams this week too and very interested to see your findings. 

Would be neat the interface for unmarshaling was similar to JSON package:

``` go
json.Unmarshal([]byte(str), &res)
ddb.Unmarshal(..., &res)
```

**Edit:** looks like this functionality already exists w/ the [dynamodbattribute](https://godoc.org/github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute#UnmarshalMap) package

``` go
e := Event{}
dynamodbattribute.UnmarshalMap(record.Dynamodb.NewImage, &e)
```

So in theory it could be used like this:

``` go
func main() {
    apexDynamo.HandleFunc(func(event *apexDynamo.Event, ctx *apex.Context) error {
        for _, record := range event.Records {
            if record.EventName == "INSERT" {
                m := Message{}
                dynamodbattribute.UnmarshalMap(record.Dynamodb.NewImage, &m)
                fmt.Println(m)
            }
        }

        return nil
    })
}
```

**Comment by rjocoleman** [comment]
Thanks @harlow 

I'm quite successfully dealing with Dynamo Events using my change from https://github.com/apex/go-apex/issues/17#issuecomment-216034101 and your pattern for `UnmarshalMap`

@tj do you want a PR for this? Dynamo Events as they stand in master are unusable unless used with a table containing those hardcoded keys.

**Comment by rjocoleman** [comment]
closed by #21
